From d567656c925204c483fd60083b57c12649b12f26 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Thu, 29 Aug 2019 12:01:03 +0800
Subject: [PATCH 25/81] modesetting: dri2: Fix memory leak of deviceName

The device name is strdup-ed, thus we need to free it in CloseScreen().

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 hw/xfree86/drivers/modesetting/dri2.c            | 6 ++++++
 hw/xfree86/drivers/modesetting/drmmode_display.h | 2 ++
 2 files changed, 8 insertions(+)

diff --git a/hw/xfree86/drivers/modesetting/dri2.c b/hw/xfree86/drivers/modesetting/dri2.c
index 62eca13c9..cc9b3544f 100644
--- a/hw/xfree86/drivers/modesetting/dri2.c
+++ b/hw/xfree86/drivers/modesetting/dri2.c
@@ -1063,6 +1063,7 @@ ms_dri2_screen_init(ScreenPtr screen)
     info.fd = ms->fd;
     info.driverName = NULL; /* Compat field, unused. */
     info.deviceName = drmGetDeviceNameFromFd(ms->fd);
+    ms->drmmode.dri2_device_name = info.deviceName;
 
     info.version = 9;
     info.CreateBuffer = ms_dri2_create_buffer;
@@ -1108,7 +1109,12 @@ ms_dri2_screen_init(ScreenPtr screen)
 void
 ms_dri2_close_screen(ScreenPtr screen)
 {
+    ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
+    modesettingPtr ms = modesettingPTR(scrn);
+
     DRI2CloseScreen(screen);
+
+    free((char *)ms->drmmode.dri2_device_name);
 }
 
 #endif /* GLAMOR_HAS_GBM */
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.h b/hw/xfree86/drivers/modesetting/drmmode_display.h
index 0db8b6f7f..59d5b19c7 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.h
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.h
@@ -145,6 +145,8 @@ typedef struct {
 
     PixmapPtr fbcon_pixmap;
 
+    const char *dri2_device_name;
+
     Bool dri2_flipping;
     Bool present_flipping;
     Bool flip_bo_import_failed;
-- 
2.38.5

